src/index_map.rs

```diff
@@ -302,43 +302,65 @@ where
     where
         F: FnMut(&mut K, &mut V) -> bool,
     {
-        const INIT: Option<Pos> = None;
-
         self.entries
             .retain_mut(|entry| keep(&mut entry.key, &mut entry.value));
 
         if self.entries.len() < self.indices.len() {
-            for index in self.indices.iter_mut() {
-                *index = INIT;
-            }
+            self.after_removal();
+        }
+    }
+
+    fn shift_remove_index(&mut self, index: usize) -> Option<(K, V)> {
+        if index >= self.entries.len() {
+            return None;
+        }
+
+        let bucket = self.entries.remove(index);
+
+        self.after_removal();
+
+        Some((bucket.key, bucket.value))
+    }
+
+    fn shift_remove_found(&mut self, _probe: usize, found: usize) -> (K, V) {
+        let entry = self.entries.remove(found);
+
+        self.after_removal(); /* Todo: pass probe if this starts taking an index parameter */
+
+        (entry.key, entry.value)
+    }
 
-            for (index, entry) in self.entries.iter().enumerate() {
-                let mut probe = entry.hash.desired_pos(Self::mask());
-                let mut dist = 0;
-
-                probe_loop!(probe < self.indices.len(), {
-                    let pos = &mut self.indices[probe];
-
-                    if let Some(pos) = *pos {
-                        let entry_hash = pos.hash();
-
-                        // robin hood: steal the spot if it's better for us
-                        let their_dist = entry_hash.probe_distance(Self::mask(), probe);
-                        if their_dist < dist {
-                            Self::insert_phase_2(
-                                &mut self.indices,
-                                probe,
-                                Pos::new(index, entry.hash),
-                            );
-                            break;
-                        }
-                    } else {
-                        *pos = Some(Pos::new(index, entry.hash));
+    // Todo: Should this take in a parameter to allow it to only process the moved
+    // elements?
```

Comment:
Yes. Technically it's *O*(n) in both cases, but one is *O*(`len()`) while the other is *O*(`len() / 2`).

Reply:
Done (see subsequent push). I don't fully understand the hashing logic, so I may have missed a subtlety.

Reply:
This is wrong, fix incoming.

Reply:
Please also add another test for the wrong logic.

Reply:
I've added a test for the wrong logic. However, I'm having a hard time figuring out what the correct logic should be.

I can replace only the affected entries' slots in `indices` with None, and and restrict the recalculation to only the affected entries. But, that fails too, because we would also need to rewrite "unaffected" entries that would have "preferred" to have a slot value that has since been vacated. (There's no longer a mismatched entry there, so when probed, it shows no value present, and the lookup fails.)

I can revert the "optimization" of `after_removal` to again re-process everything, and the test then passes. But we lose the putative optimization opportunity.

Reply:
As a test, I'm going to try the re-hashing logic from `remove_found`. It doesn't have the "robin hood" logic in it, but when I tried just leaving that logic intact I got an infinite loop. (Or I can abandon the efficiency quest for now.)

Reply:
I put the logic from `retain` back inline; I don't need to solve the problem of arbitrary removals, just a single removal. I switched to having just CoreMap::shift_remove_found, which now mirrors remove_found, plus an extra step to "shift" the appropriate index values after fixing up the removed slot. All IndexMap::shift_* operations are implemented in terms of shift_remove_found now. There's now moderate code duplication between `remove_found` and `shift_remove_found`, however.